taxcalc/parameters.py

```diff
@@ -241,9 +249,52 @@ def adjust_with_indexing(self, params_or_path, **kwargs):
             super().adjust(init_vals, **kwargs)
```

Comment:
@Peter-Metz I'm not sure why we need to reset to the initial values now that we are only deleting the auto-generated values. I changed this block to:

```
            init_vals = {}
            to_delete = {}
            for param in params:
                if param == "CPI_offset" or param in self._wage_indexed:
                    continue
                if param.endswith("-indexed"):
                    param = param.split("-indexed")[0]
                if self._data[param].get("indexed", False):
                    # init_vals[param] = pt.select_lte(
                    #     self._init_values[param],
                    #     True,
                    #     {"year": cpi_min_year["year"]},
                    # )
                    to_delete[param] = self.select_gt(
                        param, strict=True, _auto=True, year=cpi_min_year["year"]
                    )
                    needs_reset.append(param)
            self.delete(to_delete, **kwargs)
            # super().adjust(init_vals, **kwargs)
```
And all tests pass except for the `TCJA.json` reform and the round trip test: `test_round_trip_tcja_reform` in `test_reforms.py`. I'll push this branch to gh so you can see it.

Reply:
https://github.com/Peter-Metz/Tax-Calculator/compare/cpi_bug...hdoupe:init-question

Reply:
@hdoupe In theory, I agree that this would be better. Looking at the results from `test_round_trip_tcja_reform`, it looks like `2017_law.json` is correctly implemented, but the parameters aren't adjusted for `TCJA.json`. Do you think this might have something to do with how Paramtools identifies `_auto=True`? That's the only explanation that comes to mind at the moment.

Reply:
Hey @Peter-Metz sorry for the delay on replying. I worked on this a little bit, but I'm still trying to figure out what's going on with `test_round_trip_tcja_reform`. There were a couple bugs in the initial code that I shared that I've fixed here: https://github.com/Peter-Metz/Tax-Calculator/compare/cpi_bug...hdoupe:init-question

Namely:
- Fixes bug with overwriting of parameter value
- Fixes `pt.select_*` calls to get `_auto` with `select_eq` and
  then do `select_gt` query on the result.

So, the bug with `_auto` seems like it was in the code using PT but it's possible there is another bug that I haven't found yet in PT. 

I added a breakpoint in `test_round_trip_tcja_reform` so I could dig into the data some. I mostly focused on the `EITC_c` parameter:

- Values are definitely different.
  - fails: `np.testing.assert_allclose(pol._EITC_c.round(2), clp_pol._EITC_c.round(2))`
- However, it looks like the inflation rates are the same:
  - passes: np.testing.assert_allclose(pol.inflation_rates(), clp_pol.inflation_rates())
  - passes: np.testing.assert_allclose((pol._EITC_c[1:] / pol._EITC_c[:-1]).round(2), (clp_pol._EITC_c[1:] / clp_pol._EITC_c[:-1]).round(2))

I updated the [`test_round_trip_tcja_reform`](https://github.com/hdoupe/Tax-Calculator/blob/44abebd7a89e89f792e6e31d0a5583c027a08b46/taxcalc/tests/test_reforms.py#L128-L136) test with these checks on the indexing rates, too.

```
taxcalc/tests/test_reforms.py::test_round_trip_tcja_reform 
>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> PDB set_trace >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>
> /home/hankdoupe/Documents/Tax-Calculator/taxcalc/tests/test_reforms.py(129)test_round_trip_tcja_reform()
-> if fail_dump:
(Pdb) clp_pol._EITC_c # clp_pol = Policy()
array([[ 487.  , 3250.  , 5372.  , 6044.  ],
       [ 496.  , 3305.  , 5460.  , 6143.  ],
       [ 503.  , 3359.  , 5548.  , 6242.  ],
       [ 506.  , 3373.  , 5572.  , 6269.  ],
       [ 510.  , 3400.  , 5616.  , 6318.  ],
       [ 519.  , 3461.  , 5716.  , 6431.  ],
       [ 529.  , 3526.  , 5828.  , 6557.  ],
       [ 537.36, 3581.71, 5920.08, 6660.6 ],
       [ 549.07, 3659.79, 6049.14, 6805.8 ],
       [ 561.53, 3742.87, 6186.46, 6960.29],
       [ 574.5 , 3829.33, 6329.37, 7121.07],
       [ 587.25, 3914.34, 6469.88, 7279.16],
       [ 599.93, 3998.89, 6609.63, 7436.39],
       [ 612.17, 4080.47, 6744.47, 7588.09],
       [ 624.54, 4162.9 , 6880.71, 7741.37],
       [ 637.03, 4246.16, 7018.32, 7896.2 ],
       [ 649.64, 4330.23, 7157.28, 8052.54],
       [ 662.57, 4416.4 , 7299.71, 8212.79]])
(Pdb) pol._EITC_c # pol is round trip reform instance
array([[ 487.  , 3250.  , 5372.  , 6044.  ],
       [ 496.  , 3305.  , 5460.  , 6143.  ],
       [ 503.  , 3359.  , 5548.  , 6242.  ],
       [ 506.  , 3373.  , 5572.  , 6269.  ],
       [ 510.  , 3400.  , 5616.  , 6318.  ],
       [ 520.86, 3472.42, 5735.62, 6452.57],
       [ 533.62, 3557.49, 5876.14, 6610.66],
       [ 542.05, 3613.7 , 5968.98, 6715.11],
       [ 553.87, 3692.48, 6099.1 , 6861.5 ],
       [ 566.44, 3776.3 , 6237.55, 7017.26],
       [ 579.52, 3863.53, 6381.64, 7179.36],
       [ 592.39, 3949.3 , 6523.31, 7338.74],
       [ 605.19, 4034.6 , 6664.21, 7497.26],
       [ 617.54, 4116.91, 6800.16, 7650.2 ],
       [ 630.01, 4200.07, 6937.52, 7804.73],
       [ 642.61, 4284.07, 7076.27, 7960.82],
       [ 655.33, 4368.89, 7216.38, 8118.44],
       [ 668.37, 4455.83, 7359.99, 8280.  ]])
(Pdb) np.testing.assert_allclose(pol.inflation_rates(), clp_pol.inflation_rates())
(Pdb) np.allclose(pol.inflation_rates(), clp_pol.inflation_rates())
True
(Pdb) pol._EITC_c[1:] / pol._EITC_c[:-1] - 1
array([[0.01848049, 0.01692308, 0.01638124, 0.01637988],
       [0.0141129 , 0.01633888, 0.01611722, 0.0161159 ],
       [0.00596421, 0.00416791, 0.00432588, 0.00432554],
       [0.00790514, 0.00800474, 0.00789663, 0.00781624],
       [0.02129412, 0.0213    , 0.02129986, 0.02129946],
       [0.02449795, 0.02449876, 0.02449953, 0.02450032],
       [0.01579776, 0.01580047, 0.01579949, 0.01580024],
       [0.02180611, 0.02180037, 0.02179937, 0.02180009],
       [0.02269486, 0.02270019, 0.02270007, 0.02270058],
       [0.02309159, 0.02309933, 0.02310042, 0.02310018],
       [0.02220803, 0.02219991, 0.02219962, 0.02219975],
       [0.02160739, 0.02159876, 0.02159946, 0.02160044],
       [0.02040681, 0.02040103, 0.02040002, 0.02039945],
       [0.02019302, 0.02019962, 0.02019952, 0.02019947],
       [0.01999968, 0.01999967, 0.01999994, 0.01999941],
       [0.01979428, 0.01979893, 0.01979998, 0.01979947],
       [0.01989837, 0.01989979, 0.01990056, 0.01990037]])
(Pdb) pol.inflation_rates()
[0.0148, 0.0159, 0.0012, 0.0126, 0.0188, 0.022000000000000002, 0.0158, 0.0218, 0.0227, 0.023100000000000002, 0.0222, 0.0216, 0.0204, 0.020200000000000003, 0.02, 0.0198, 0.0199, 0.0198]
(Pdb) clp_pol._EITC_c[1:] / clp_pol._EITC_c[:-1] - 1
array([[0.01848049, 0.01692308, 0.01638124, 0.01637988],
       [0.0141129 , 0.01633888, 0.01611722, 0.0161159 ],
       [0.00596421, 0.00416791, 0.00432588, 0.00432554],
       [0.00790514, 0.00800474, 0.00789663, 0.00781624],
       [0.01764706, 0.01794118, 0.01780627, 0.01788541],
       [0.01926782, 0.0187807 , 0.01959412, 0.0195926 ],
       [0.0158034 , 0.01579977, 0.01579959, 0.01579991],
       [0.02179172, 0.02179964, 0.02180038, 0.02179984],
       [0.02269292, 0.02270076, 0.02270075, 0.02269976],
       [0.02309761, 0.02309992, 0.02310045, 0.02309961],
       [0.02219321, 0.02219971, 0.02219968, 0.02220032],
       [0.02159217, 0.02160007, 0.02160009, 0.02160002],
       [0.02040238, 0.02040066, 0.02040054, 0.02039968],
       [0.02020681, 0.0202011 , 0.02020025, 0.02020008],
       [0.01999872, 0.02000048, 0.01999939, 0.02000034],
       [0.01979499, 0.01979907, 0.01979961, 0.0197994 ],
       [0.01990333, 0.01989964, 0.01990002, 0.01990055]])
(Pdb) np.testing.assert_allclose((pol._EITC_c[1:] / pol._EITC_c[:-1]).round(2), (clp_pol._EITC_c[1:] / clp_pol._EITC_c[:-1]).round(2))
(Pdb) np.testing.assert_allclose(pol._EITC_c.round(2), clp_pol._EITC_c.round(2))
*** AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 52 / 72 (72.2%)
Max absolute difference: 67.21
Max relative difference: 0.00893221
 x: array([[ 487.  , 3250.  , 5372.  , 6044.  ],
       [ 496.  , 3305.  , 5460.  , 6143.  ],
       [ 503.  , 3359.  , 5548.  , 6242.  ],...
 y: array([[ 487.  , 3250.  , 5372.  , 6044.  ],
       [ 496.  , 3305.  , 5460.  , 6143.  ],
       [ 503.  , 3359.  , 5548.  , 6242.  ],...
(Pdb) 
```

Reply:
Thanks a ton for the review @hdoupe and I appreciate your improvements!  One initial thought, and then I'll also dig into this more later. I agree that the `inflation_rates()` are the same, but `._EITC_c[1:] / ._EITC_c[:-1]` are not the same beyond two decimal points:

```
(Pdb) np.testing.assert_allclose((pol._EITC_c[1:] / pol._EITC_c[:-1]).round(4), (clp_pol._EITC_c[1:] / clp_pol._EITC_c[:-1]).round(4))
*** AssertionError: 
Not equal to tolerance rtol=1e-07, atol=0

Mismatched elements: 8 / 68 (11.8%)
Max absolute difference: 0.0057
Max relative difference: 0.00559482
 x: array([[1.0185, 1.0169, 1.0164, 1.0164],
       [1.0141, 1.0163, 1.0161, 1.0161],
       [1.006 , 1.0042, 1.0043, 1.0043],...
 y: array([[1.0185, 1.0169, 1.0164, 1.0164],
       [1.0141, 1.0163, 1.0161, 1.0161],
       [1.006 , 1.0042, 1.0043, 1.0043],...
```

Reply:
Oh @Peter-Metz you're totally right. I was wondering how that could be possible....

Reply:
Thanks for reviewing. Let me know if you have any other ideas.